xla/mlir/backends/openxla/conversion/convert_memref_ops.cc

```diff
@@ -0,0 +1,254 @@
+// no corresponding operation in tensor/flow/hal dialect that allows type
+// conversion, so we implement "tensor.view" through conversion to HAL buffers.
+FailureOr<IREE::Input::TensorImportOp> reinterpretCastTensor(
+    ImplicitLocOpBuilder &b, TypeConverter &converter, Value source,
+    Value offset, MemRefType memref_ty, NotifyMatchFailure match_failure) {
+  auto tensor_ty = converter.convertType(memref_ty).cast<RankedTensorType>();
+
+  // Export tensor as !iree_input.buffer.
+  auto buffer_ty = b.getType<IREE::Input::BufferType>();
+  auto export_op = b.create<IREE::Input::TensorExportOp>(
+      buffer_ty, source, /*source_dims=*/ValueRange());
+
+  // Element type must be supported by the IREE HAL ABI.
+  auto abi_ty = IREE::Input::getElementTypeValue(tensor_ty.getElementType());
+  if (!abi_ty.has_value()) return match_failure("unsupported element type");
+
+  // Create a buffer view with given offset, type and shape.
+  auto buffer_view_size = std::max(1u, tensor_ty.getElementTypeBitWidth() / 8) *
+                          tensor_ty.getNumElements();
+
+  auto buffer_view_shape = llvm::to_vector(
+      llvm::map_range(tensor_ty.getShape(), [&](int64_t dim) -> Value {
+        return b.create<arith::ConstantIndexOp>(dim);
+      }));
+
+  auto view_op = b.create<IREE::Input::BufferViewCreateOp>(
+      export_op, offset, b.create<arith::ConstantIndexOp>(buffer_view_size),
+      *abi_ty,
+      /*IREE_HAL_ENCODING_TYPE_DENSE_ROW_MAJOR*/ 1, buffer_view_shape);
+
+  // Import it back as a properly typed tensor.
+  return b.create<IREE::Input::TensorImportOp>(tensor_ty, view_op.getResult(),
+                                               /*target_dims=*/ValueRange());
+}
+
+//===----------------------------------------------------------------------===//
+// Converts memref.view op to a iree_input.tensor.{export,import}
+//===----------------------------------------------------------------------===//
+
+struct ConvertMemrefViewOp : public OpConversionPattern<memref::ViewOp> {
+  ConvertMemrefViewOp(TypeConverter &converter, MLIRContext *ctx,
+                      std::shared_ptr<DeBufferization> state)
+      : OpConversionPattern(converter, ctx), state(std::move(state)) {}
+
+  LogicalResult matchAndRewrite(
+      memref::ViewOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    ImplicitLocOpBuilder b(op.getLoc(), rewriter);
+
+    NotifyMatchFailure match_failure = [&](const char *msg) {
+      return rewriter.notifyMatchFailure(op, msg);
+    };
+
+    auto memref = op.getType().dyn_cast<MemRefType>();
+    if (!memref)
+      return rewriter.notifyMatchFailure(op, "expected a memref result");
+
+    auto source = adaptor.getSource().cast<BlockArgument>();
+
+    // TODO(ezhulenev): Is it really needed? Try to delete it.
+    auto barrier =
+        b.create<IREE::Input::OptimizationBarrierOp>(source).getResult(0);
```

Comment:
I'd be curious about the results of any investigation here. Most of our uses of `OptimizationBarrierOp` are in tests

* https://openxla.github.io/iree/reference/mlir-dialects/IREEInput/#iree_inputoptimization_barrier-inputoptimizationbarrierop
* https://github.com/search?q=repo%3Aopenxla%2Firee%20optimization_barrier&type=code

Reply:
I'm pretty sure this one is not needed, it's a leftover from early experiments. Will try to remove it in next PRs. The only optimization barriers that are needed, are right before the `return` to prevent DCE